File: frontend/src/lib/api/activity-logs.ts
import { apiClient } from '@/lib/api-client'

export interface ActivityLog {
  id: number
  user_id: number | null
  user_email: string | null
  user_name: string | null
  action: string
  resource_type: string
  resource_id: string | null
  http_method: string | null
  endpoint: string | null
  ip_address: string | null
  response_status: number | null
  response_time_ms: number | null
  details: Record<string, unknown> | null
  created_at: string
}

export interface ActivityLogStats {
  total_requests: number
  total_errors: number
  error_rate: number
  avg_response_time_ms: number
  requests_by_action: Record<string, number>
  requests_by_resource: Record<string, number>
  requests_by_user: Array<{
    user_id: number
    email: string
    name: string
    count: number
  }>
  requests_over_time: Array<{
    period: string
    count: number
  }>
}

export interface ActivityLogListResponse {
  items: ActivityLog[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface ActivityLogListParams {
  page?: number
  page_size?: number
  user_id?: number
  action?: string
  resource_type?: string
  response_status?: number
  date_from?: string
  date_to?: string
  search?: string
  sort_by?: string
  sort_order?: 'asc' | 'desc'
}

export const activityLogsApi = {
  list: async (params?: ActivityLogListParams): Promise<ActivityLogListResponse> => {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null && value !== '') {
          queryParams.append(key, value.toString())
        }
      })
    }
    const queryString = queryParams.toString()
    const url = queryString ? `/v1/activity-logs/?${queryString}` : '/v1/activity-logs/'
    const response = await apiClient.get(url)
    return response.data
  },

  getStats: async (dateFrom?: string, dateTo?: string): Promise<ActivityLogStats> => {
    const params = new URLSearchParams()
    if (dateFrom) params.append('date_from', dateFrom)
    if (dateTo) params.append('date_to', dateTo)
    const queryString = params.toString()
    const url = queryString ? `/v1/activity-logs/stats?${queryString}` : '/v1/activity-logs/stats'
    const response = await apiClient.get(url)
    return response.data
  },

  getActions: async (): Promise<string[]> => {
    const response = await apiClient.get('/v1/activity-logs/actions')
    return response.data
  },

  getResourceTypes: async (): Promise<string[]> => {
    const response = await apiClient.get('/v1/activity-logs/resource-types')
    return response.data
  },

  exportCsv: async (params?: ActivityLogListParams): Promise<void> => {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null && value !== '') {
          queryParams.append(key, value.toString())
        }
      })
    }
    const queryString = queryParams.toString()
    const url = queryString ? `/v1/activity-logs/export?${queryString}` : '/v1/activity-logs/export'

    const response = await apiClient.get(url, { responseType: 'blob' })
    const blob = new Blob([response.data], { type: 'text/csv;charset=utf-8' })
    const downloadUrl = URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = downloadUrl
    a.download = `aktivitetslogg-${new Date().toISOString().split('T')[0]}.csv`
    a.click()
    URL.revokeObjectURL(downloadUrl)
  },

  cleanupOldLogs: async (daysToKeep: number = 90): Promise<{ message: string }> => {
    const response = await apiClient.post(`/v1/activity-logs/cleanup?days_to_keep=${daysToKeep}`)
    return response.data
  },
}

File: backend/app/api/v1/activity_logs.py
"""Activity log API endpoints."""
from datetime import datetime
from typing import Optional, Literal, List

from fastapi import APIRouter, Depends, Query, HTTPException
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_active_user, get_db
from app.domain.entities.user import User
from app.services.activity_log_service import ActivityLogService
from app.schemas.activity_log import (
    ActivityLog,
    ActivityLogListResponse,
    ActivityLogStats,
)

router = APIRouter()


def require_admin(current_user: User) -> User:
    """Check if current user is admin."""
    if current_user.rolle != "admin" and not getattr(current_user, 'is_superuser', False):
        raise HTTPException(
            status_code=403,
            detail="Kun administratorer kan se aktivitetsloggen"
        )
    return current_user


@router.get("/", response_model=ActivityLogListResponse)
async def get_activity_logs(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=100),
    user_id: Optional[int] = Query(None),
    action: Optional[str] = Query(None),
    resource_type: Optional[str] = Query(None),
    response_status: Optional[int] = Query(None),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    search: Optional[str] = Query(None, max_length=200),
    sort_by: str = Query("created_at"),
    sort_order: Literal["asc", "desc"] = Query("desc"),
):
    """Get activity logs with pagination and filters. Admin only."""
    require_admin(current_user)

    service = ActivityLogService(db)
    items, total = await service.list_logs(
        page=page,
        page_size=page_size,
        user_id=user_id,
        action=action,
        resource_type=resource_type,
        response_status=response_status,
        date_from=date_from,
        date_to=date_to,
        search=search,
        sort_by=sort_by,
        sort_order=sort_order,
    )

    total_pages = (total + page_size - 1) // page_size

    return ActivityLogListResponse(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        total_pages=total_pages,
    )


@router.get("/stats", response_model=ActivityLogStats)
async def get_activity_stats(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
):
    """Get activity log statistics. Admin only."""
    require_admin(current_user)

    service = ActivityLogService(db)
    return await service.get_stats(date_from=date_from, date_to=date_to)


@router.get("/export")
async def export_activity_logs(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
    user_id: Optional[int] = Query(None),
    action: Optional[str] = Query(None),
    resource_type: Optional[str] = Query(None),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
):
    """Export activity logs to CSV. Admin only."""
    require_admin(current_user)

    service = ActivityLogService(db)
    csv_content = await service.export_to_csv(
        user_id=user_id,
        action=action,
        resource_type=resource_type,
        date_from=date_from,
        date_to=date_to,
    )

    filename = f"aktivitetslogg-{datetime.now().strftime('%Y%m%d-%H%M%S')}.csv"

    return StreamingResponse(
        iter(['\ufeff' + csv_content]),  # BOM for Excel compatibility
        media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": f"attachment; filename={filename}"}
    )


@router.get("/actions", response_model=List[str])
async def get_available_actions(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    """Get list of available action types from logs."""
    require_admin(current_user)
    service = ActivityLogService(db)
    return await service.get_unique_actions()


@router.get("/resource-types", response_model=List[str])
async def get_resource_types(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
):
    """Get list of unique resource types from logs."""
    require_admin(current_user)
    service = ActivityLogService(db)
    return await service.get_unique_resource_types()


@router.post("/cleanup")
async def cleanup_old_logs(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
    days_to_keep: int = Query(90, ge=30, le=365),
):
    """Delete old activity logs. Superuser only."""
    require_admin(current_user)

    if not getattr(current_user, 'is_superuser', False):
        raise HTTPException(
            status_code=403,
            detail="Kun superbrukere kan slette aktivitetslogger"
        )

    service = ActivityLogService(db)
    deleted_count = await service.cleanup_old_logs(days_to_keep)

    return {"message": f"Slettet {deleted_count} gamle loggoppforinger"}
